articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,230 @@
+
+The typical approaches to synchronizing concurrent access to data tend to decrease the determinism of a system and increase the complexity.
+For example, mutexes introduce the possibility of deadlock when used improperly.
+As mentioned above, many real-time operating systems offer alternative synchronization primitives for real-time safety.
+
+Guidelines for synchronization in ROS 2:
+
+- Prefer atomics to mutexes when possible.
+
+- Thoroughly test multithreaded code for deadlock and livelock at the earliest opportunity.
+
+  - Consider compiling multithreaded regression tests with the Clang ThreadSanitizer (3).
+
+- Hide synchronization primitives behind a layer of abstraction that wraps standard library classes such as `std::mutex` and `std::condition_variable` as the default behavior.
+
+  - Inject alternative synchronization primitives from RTOS-specific APIs where applicable.
+
+As of the time of writing, the Executor and the IntraProcessManager in rclcpp are designed to partially follow the last guideline.
+However, the abstractions are messy (particularly for the IntraProcessManager).
+There is no real-time multithreaded Executor that uses "real-time safe" synchronization primitives, similarly there is no such IntraProcessManager--there is only the abstraction path that would allow such alternatives to be implemented.
+There are also synchronization primitives that are not hidden behind an abstraction layer for the global interrupt handler and for parameters.
+
+### Middleware selection and QoS
+
+If the underlying middleware used in ROS 2 is not suitable for real-time performance, the entire system will fail to meet the real-time requirement.
+Network communication and discovery is often a source of nondeterminism.
+Though some real-time operating systems partially mitigate this (e.g. disabling interrupts), careful selection and use of a middleware is important.
+
+The embedded component of the system proposed will likely use `freertps` as the underlying middleware.
+The "desktop" system can be matrix-tested using the various DDS implementations supported by ROS 2.
+This allows us to compare the real-time performance of the different DDS implementations.
+
+Quality of service is always an important consideration with DDS.
+Some DDS QoS options were designed with real-time performance in mind, such as the DEADLINE policy (4).
+It may be convenient to expose these policies and provide a "real-time" QoS profile in rmw as a suggestion to ROS 2 users.
+Benchmarking and profiling may provide data to support the introduction of more QoS options into the rmw API.
+
+### Component lifecycle
+
+It should be noted that the planned component lifecycle features are useful for real-time programming, since it is common that real-time systems have a initialization phase in which the real-time constraint is relaxed, in which it is safe to preallocate memory, change thread priority, etc.
```

Comment:
The reference was added in 928dd5f9153b790b0180d976d1bfde4df4333bda

https://github.com/ros2/design/commit/928dd5f9153b790b0180d976d1bfde4df4333bda#diff-2309eff1b7fac7aaf93df0fde7d4c62eR155